components/masking_functions/src/udf/udf_registration.cc

```diff
@@ -0,0 +1,900 @@
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software Foundation,
+   51 Franklin Street, Suite 500, Boston, MA 02110-1335 USA */
+
+#include <array>
+#include <bitset>
+
+#include "mysql/components/component_implementation.h"
+#include "mysql/components/my_service.h"
+#include "mysql/components/services/registry.h"
+#include "mysql/components/services/udf_registration.h"
+
+#include "components/masking_functions/include/component.h"
+#include "components/masking_functions/include/udf/udf_registration.h"
+#include "components/masking_functions/include/udf/udf_utils.h"
+#include "components/masking_functions/include/udf/udf_utils_sql.h"
+#include "components/masking_functions/include/udf/udf_utils_string.h"
+
+#include <mysql/components/services/udf_metadata.h>
+extern REQUIRES_SERVICE_PLACEHOLDER(mysql_udf_metadata);
+extern REQUIRES_SERVICE_PLACEHOLDER(udf_registration);
+extern REQUIRES_SERVICE_PLACEHOLDER(mysql_string_character_access);
+extern REQUIRES_SERVICE_PLACEHOLDER(mysql_string_converter);
+extern REQUIRES_SERVICE_PLACEHOLDER(mysql_string_factory);
+
+#define MYSQLPP_CHARSET_SUPPORT
+#include <mysqlpp/udf_wrappers.hpp>
+
+class gen_range_impl {
+ public:
+  gen_range_impl(mysqlpp::udf_context &ctx) {
+    if (ctx.get_number_of_args() != 2)
+      throw std::invalid_argument("Wrong argument list: should be (int, int)");
+
+    ctx.mark_result_nullable(true);
+    ctx.mark_result_const(false);
+
+    ctx.mark_arg_nullable(0, false);
+    ctx.set_arg_type(0, INT_RESULT);
+
+    ctx.mark_arg_nullable(1, false);
+    ctx.set_arg_type(1, INT_RESULT);
+  }
+
+  mysqlpp::udf_result_t<INT_RESULT> calculate(const mysqlpp::udf_context &ctx) {
+    const long long lower = *ctx.get_arg<INT_RESULT>(0);
+    const long long upper = *ctx.get_arg<INT_RESULT>(1);
+
+    if (upper < lower) {
+      return std::nullopt;
+    } else {
+      return mysql::plugins::random_number(lower, upper);
+    }
+  }
+};
+
+class gen_rnd_email_impl {
+ public:
+  gen_rnd_email_impl(mysqlpp::udf_context &ctx) {
+    if (ctx.get_number_of_args() > 2)
+      throw std::invalid_argument(
+          "Wrong argument list: should be ([int], [string])");
+
+    ctx.mark_result_nullable(true);
+    ctx.mark_result_const(false);
+
+    if (ctx.get_number_of_args() >= 1) {
+      ctx.mark_arg_nullable(0, false);
+      ctx.set_arg_type(0, INT_RESULT);
+    }
+
+    if (ctx.get_number_of_args() >= 2) {
+      ctx.mark_arg_nullable(1, false);
+      ctx.set_arg_type(1, STRING_RESULT);
+    }
+
+    if (ctx.get_number_of_args() >= 2) {
+      ctx.set_return_value_charset(ctx.get_arg_charset(1));
```

Comment:
Didn't you created a helper function for this?

Reply:
I did, I changed it to use it where possible.